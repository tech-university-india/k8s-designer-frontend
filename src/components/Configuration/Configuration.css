
.config-layout{
    margin: auto;
    margin-top: 10px;
    display: flex;
    flex-direction: column;
    align-items: center;
}
.config-layout-container{
    margin: auto;
    margin-top: 10px;
    display: flex;
    flex-direction: column;
    align-items: center;
    overflow: auto;
    position: relative;

}
.config-btn-color{
    margin:10px auto 10px auto!important;
    background-color: rgb(19, 51, 124) !important;
    color: white !important;
    width: 18.2vw;
    height: 7vh;
}
.add-textfield-container{
    position: absolute;
    right: 12px;
    bottom: 247px;
    color: rgb(19, 51, 124)!important;
}

.service-element-border{
    height: 5vh;
    font-size: 20px;
    border-bottom : 2px solid rgb(82, 81, 81);
    text-align: center;
    padding-top: 2vh;
    
}

@media screen and (max-width:768px) {
    .service-element-border{
        font-size: 15px;
        height: 7vh;
        padding-top: 1.5vh;

    }
    .config-layout>{
        font-size: 15px!important;
    }
    .add-textfield-container{
        right: -5px;
        bottom: 270px;
    }
    
}