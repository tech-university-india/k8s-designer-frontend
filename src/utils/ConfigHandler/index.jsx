const ConfigHandler = (event,updateConfig,userFeConfig,setUserFeConfig)=>{
  setUserFeConfig({
    ...userFeConfig,
    [updateConfig]:event.target.value
  });
};

export default ConfigHandler;