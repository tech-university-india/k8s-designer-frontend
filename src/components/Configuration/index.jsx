import * as React from 'react';
import Box from '@mui/material/Box';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import AddCircleOutlineIcon from '@mui/icons-material/AddCircleOutline';
import './Configuration.css';
import PropTypes from 'prop-types';
import ConfigHandler from '../../utils/ConfigHandler';
import {CustomTextField} from '../index';
import { REPOSITORY , getConfiguration } from '../../utils/makeRequest';


export default function Configuration({service}){

  

  const [userConfig , setUserConfig] = React.useState();
  const [userRepoConfig , setUserRepoConfig] = React.useState();

  const [newTextField, setNewTextField] = React.useState([]);

  React.useEffect(() => {
    setUserRepoConfig(REPOSITORY);
    setUserConfig(getConfiguration(service));
  },[service]);

  // React.useEffect(() => {
  //   return()=>{

  //     // updateConfiguration(service,userRepoConfig);
  //     // updateConfiguration(userRepoConfig);
  //   };
  // },[userConfig,userRepoConfig]);

  const handleUserConfig = (newUserConfig)=>{
    setUserConfig(newUserConfig);
  };
   
  const onAddBtnClick = () => {
    setNewTextField((<CustomTextField userConfig = {userConfig} handleUserConfig={handleUserConfig} setNewTextField = {setNewTextField}/>));
  };

  return userConfig ?(
    <div>
      <div className='service-element-border'>{service} MICROSERVICES</div>
      <div  className='config-layout-container'>
        <Box
          component="form"
          sx={{
            '& > :not(style)': { m: 1, width: '18.2vw' },
          }}
          noValidate
          autoComplete="off"
          className='config-layout'
          style = {
            { fontFamily :  'Open Sans'}
          }
        >
          {Object.keys(userConfig).map((configuration,index)=>(
            <TextField id="outlined-basic" label= {`Enter ${configuration}`} value = {userConfig[configuration] || ''} key={index} onChange = {(event)=>ConfigHandler(event,configuration,userConfig,setUserConfig)} variant="outlined" />
          ))}
          {newTextField}
          {Object.keys(userRepoConfig).map((configuration,index)=>(
            <TextField id="outlined-basic" label= {`Enter ${configuration}`} key={index}  value ={userRepoConfig[configuration]||''} onChange = {(event)=>ConfigHandler(event,configuration,userRepoConfig,setUserRepoConfig)} variant="outlined" />
          ))}
        </Box>
        {newTextField.length===0 && <div className='add-textfield-container'>
          <AddCircleOutlineIcon data-testid="add-new-config" color='blue' onClick={onAddBtnClick}/>
        </div>}
        <div className='generate-btn-container'>
          <Button className='config-btn-color'>GENERATE NOW</Button>
        </div>
      </div>
    </div>
  ) : (
    <div>
      <p>Loading...</p>
    </div>
  );
}

Configuration.propTypes = {
  service : PropTypes.string,
  userRepo:  PropTypes.object,
  userRepoHandler: PropTypes.func,
  userConfige:  PropTypes.object,
  userConfigHandler: PropTypes.func
};