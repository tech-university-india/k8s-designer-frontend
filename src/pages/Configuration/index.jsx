import './Configuration.css';
import * as React from 'react';
import {Header,FrontendConfig} from '../../components';

export default function configuration() {

  const userRepoId ={
    serverAddress : '',
    token:''
  };
  //this state should be in microservices drag and drop  component
  // const [service,setService] = React.useState('frontend');
  const [userRepo , setUserRepo] = React.useState(userRepoId);
  const [feConfig,setFeConfig] = React.useState();
  // const [beConfig,setBeConfig] = React.useState();
  // const [dbConfig,setDbConfig] = React.useState();

  const userRepoHandler = (userRepo)=>{
    setUserRepo(userRepo);
  };

  const feConfigHandler = (feConfig)=>{
    setFeConfig(feConfig);
  };
  

  return (
    <div>
      <Header /> 
      <div className='config-body'>
        <div className='config-element-border'>CONFIGURATION</div>
        <FrontendConfig userRepo = {userRepo} userRepoHandler = {userRepoHandler} feConfig = {feConfig} feConfigHandler = {feConfigHandler} />
      </div>
    </div>
    
  );
}