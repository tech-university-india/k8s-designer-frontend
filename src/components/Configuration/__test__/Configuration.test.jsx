import {fireEvent, render ,screen,waitFor } from '@testing-library/react';
import Configuration from '..';
import * as React from 'react';
import ConfigHandler from '../../../utils/ConfigHandler';


jest.mock('../../../utils/ConfigHandler');


describe('Frontend Configuration',()=>{

  it('should show  Frontend configuration  text when component is rendered',()=>{
    expect(screen.queryByText('FRONTEND MICROSERVICES')).toBeFalsy();
    render(<Configuration service = "FRONTEND"/>);
    expect(screen.queryByText('FRONTEND MICROSERVICES')).toBeTruthy();
  });
   
  

  it('should render all configuration fields when component is rendered',async()=>{
    expect(screen.queryByLabelText('Enter PORT')).toBeFalsy();
    render(<Configuration service = "FRONTEND"/>);
    await waitFor(()=>{
      expect(screen.queryByLabelText('Enter DATABASE_URL')).toBeTruthy();
    });
    expect(screen.queryByLabelText('Enter PORT')).toBeTruthy();
    expect(screen.queryByLabelText('Enter ENVIRONMENT')).toBeTruthy();
    expect(screen.queryByLabelText('Enter NUMBER_OF_REPLICAS')).toBeTruthy();
  });
  it('should call FE config handler with port as parameter when port textfield is updated',async()=>{

    render(<Configuration service = "FRONTEND"/>);
    await waitFor(()=>{
      expect(screen.queryByLabelText('Enter DATABASE_URL')).toBeTruthy();
    });
    expect(screen.queryByLabelText('Enter PORT')).toBeTruthy();
    expect(ConfigHandler).toHaveBeenCalledTimes(0);
    const PortTextfield = screen.queryByLabelText('Enter PORT');
    fireEvent.change( PortTextfield , {target : {value :'8000'}});
    expect(ConfigHandler).toHaveBeenCalledTimes(1);
    expect(ConfigHandler).toHaveBeenCalledWith(expect.anything(),'PORT',expect.anything(),expect.anything());
  });

  it('it should add new textfield when plus button is clicked',async()=>{
    render(<Configuration service = "FRONTEND"/>);
    await waitFor(()=>{
      expect(screen.getByTestId('add-new-config')).toBeTruthy();
    });
    const addButton = screen.getByTestId('add-new-config');
    fireEvent.click(addButton);
    await waitFor(()=>{
      expect(screen.getByTestId('newConfig-key')).toBeTruthy();
    });
  });

  it('it should update service configuration with new textfield when plus button is clicked',async()=>{
    render(<Configuration service = "FRONTEND"/>);
    await waitFor(()=>{
      expect(screen.getByTestId('add-new-config')).toBeTruthy();
    });
    let addButton = screen.getByTestId('add-new-config');
    fireEvent.click(addButton);
    await waitFor(()=>{
      expect(screen.getByTestId('newConfig-key')).toBeTruthy();
    });
    const configKey = screen.getByTestId('newConfig-key').querySelector('input');
    fireEvent.change(configKey,{target:{value:'NAME'}});
    const configValue = screen.queryByTestId('newConfig-value').querySelector('input');
    fireEvent.change(configValue,{target:{value:'value'}});
    const addButton2= screen.getByTestId('add-newConfig');
    fireEvent.click(addButton2);
    await waitFor(()=>{
      expect(screen.getByDisplayValue('value')).toBeTruthy();
    });
  });
  it('should render correctly and create a snapshot', () => {
    const {asFragment } =  render(<Configuration service = "FRONTEND"/>);
    expect(asFragment()).toMatchSnapshot();
  });
});