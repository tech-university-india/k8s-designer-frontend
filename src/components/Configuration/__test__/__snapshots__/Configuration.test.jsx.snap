// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frontend Configuration should render correctly and create a snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="service-element-border"
    >
      FRONTEND MICROSERVICES
    </div>
    <div
      class="config-layout-container"
    >
      <form
        autocomplete="off"
        class="config-layout MuiBox-root css-198ak82"
        novalidate=""
        style="font-family: Open Sans;"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Enter PORT
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Enter PORT
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Enter NUMBER_OF_REPLICAS
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Enter NUMBER_OF_REPLICAS
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Enter ENVIRONMENT
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Enter ENVIRONMENT
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Enter DATABASE_URL
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Enter DATABASE_URL
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Enter SERVER_ADDRESS
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Enter SERVER_ADDRESS
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Enter TOKEN
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Enter TOKEN
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </form>
      <div
        class="add-textfield-container"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorBlue MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="add-new-config"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </div>
      <div
        class="generate-btn-container"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium config-btn-color css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          GENERATE NOW
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
